### PR TITLE
Fix Exception Handling in SystemNetHttpClient + Unit test http clients

### PR DESCRIPTION
**Pull request by efossier**
Current behavior:
- SystemNetHttpHandler incorrectly assumes that SendAsync will throw an exception on a non-2xx status code, in fact it will only throw an exception on network level non-http errors that occur. This is not the end of the world because we will return a (Twilio) Response as usual and the calling code in TwilioRestClient will check the status code and convert to an Exception on an http error.
- We incorrectly catch AggregateException in MakeRequestAysnc, await only throws one exception. Calling task.Wait() will return an AggregateException. source: http://stiller.blog/en/2012/12/accessing-the-aggregateexception-with-await/

Changes:
- Add support for LastRequest, LastResponse
- Cleanup and centralize handling Non 2xx status codes into TwilioRestClient
- Always rethrow caught exceptions in SystemNetHttpClient
- Fix typo in HttpClient `MakeRequestAysnc` -> `MakeRequestAsync`
- Unpack the exception in `MakeRequest` always throw the first exception in AggregateException
- Abstract WebRequest stuff in WebRequestClient for unit testing, make exception handling/behavior cleaner and more consistent with SystemNetHttpClient
- We should be throwing more granular exceptions now than we used to, old clients used to throw a lot of generic TwilioExceptions and actual exception will get lost. We are now properly saving the root cause exception and throwing more granular exceptions such as ApiConnectionException if its a connection error, RestException in the case of a 500 etc.

### CODE REVIEW

File: test/Twilio.Test/Http/SystemNetHttpClientTest.cs [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+using System.Threading;
+using System.Threading.Tasks;
+using Castle.Core.Logging;
+using NUnit.Framework;
+using Twilio.Http;
+using HttpMethod = Twilio.Http.HttpMethod;
+
+namespace Twilio.Tests.Http
+{
+    public class MockResponseHandler : DelegatingHandler
```
Comment (efossier) [inline test/Twilio.Test/Http/SystemNetHttpClientTest.cs]: Had to add this to mock out the actual http calls in `System.Net.Http.HttpClient`. Dont think we'll use it in other tests so i left it here but we can move it to its own file anyone cares enough.

Comment (dprothero) [inline test/Twilio.Test/Http/SystemNetHttpClientTest.cs]: Wrap this whole file with `#if !NET35` (just like `src/Twilio/Http/SystemNetHttpClient.cs` is now) -- .NET 3.5 build is failing on AppVeyor.

**Comment by dprothero** [comment]
Added a comment on why AppVeyor is failing. Otherwise, this looks great to me!

File: src/Twilio/Clients/TwilioRestClient.cs [hunk 2]
```diff
@@ -67,7 +69,18 @@ public class TwilioRestClient : ITwilioRestClient
         public Response Request(Request request)
         {
             request.SetAuth(_username, _password);
-            var response = HttpClient.MakeRequest(request);
+            Response response;
+            try
+            {
+                response = HttpClient.MakeRequest(request);
+            }
+            catch (Exception innerException)
```
Comment (dougblack) [inline src/Twilio/Clients/TwilioRestClient.cs]: Why `innerException`?

File: src/Twilio/Clients/TwilioRestClient.cs [hunk 3]
```diff
@@ -99,19 +124,25 @@ private static Response ProcessResponse(Response response)
...
+            try
+            {
+                restException = RestException.FromJson(response.Content);
+            }
+            catch (JsonReaderException) { /* Allow null check below to handle */ }
+            
             if (restException == null)
             {
-                throw new ApiException("Server Error, no content");
+                throw new ApiException("Api Error: " + response.StatusCode + " - " + (response.Content ?? "[no content]"));
```
Comment (dougblack) [inline src/Twilio/Clients/TwilioRestClient.cs]: I actually like `Server Error` here. Thoughts?

Comment (dougblack) [inline src/Twilio/Clients/TwilioRestClient.cs]: Nevermind. `API Error` is fine.

File: src/Twilio/Http/HttpClient.cs [hunk 4]
```diff
@@ -8,21 +8,33 @@ namespace Twilio.Http
...
         ///
         /// <param name="request">request to make</param>
+        /// <exception>throws exception on network or connection errors.</exception>
         /// <returns>response of the request</returns>
         public abstract Response MakeRequest(Request request);
 
 #if !NET35
         /// <summary>
-        /// Make an async request to Twilio
+        /// Make an async request to Twilio, this should return a Response regardless of response status code.
```
Comment (dougblack) [inline src/Twilio/Http/HttpClient.cs]: Let's clarify that this returns non-2XX response as well.

File: src/Twilio/Clients/TwilioRestClient.cs (hunk 2 above)
Comment (efossier) [inline src/Twilio/Clients/TwilioRestClient.cs]: not really any reason. I'll rename to `clientException`?

File: src/Twilio/Http/Net35/WebRequestWrappers.cs [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+namespace Twilio.Http.Net35
+{
+    /// <summary>
+    /// These classes exist to wrap MSFT's WebRequest interfaces to enable unit testing of 
+    /// WebRequestClient.
+    /// </summary>
+    
+    public class HttpWebRequestFactory
+    {
+        public virtual IHttpWebRequestWrapper Create(Uri uri)
```
Comment (efossier) [inline src/Twilio/Http/Net35/WebRequestWrappers.cs]: ℹ️  virtual means the method can be overriden which is required for creating a mock of it.

File: src/Twilio/Http/SystemNetHttpClient.cs [hunk 6]
```diff
@@ -36,62 +32,41 @@ public SystemNetHttpClient(System.Net.Http.HttpClient httpClient = null)
...
+            {
+                var task = MakeRequestAsync(request);
+                task.Wait();
+                return task.Result;
+            }
+            catch (AggregateException ae)
+            {
+                // Combine nested AggregateExceptions
+                ae = ae.Flatten();
+                throw (ae.InnerExceptions.Count == 1) ? ae.InnerException : ae;
```
Comment (efossier) [inline src/Twilio/Http/SystemNetHttpClient.cs]: ⚠️  *Question*: Should we always return the first exception? Cant think of a scenario when wed have multiple exceptions in the AggregateException but its technically possible. Throwing an AggregateException leaks the implementation detail that the synchronous MakeRequest calls the async version under the covers.

File: src/Twilio/Http/HttpClient.cs [hunk 7]
```diff
@@ -8,21 +8,33 @@ namespace Twilio.Http
...
         /// <summary>
-        /// Make an async request to Twilio
+        /// Make an async request to Twilio, returns non-2XX responses as well
         /// </summary>
         ///
         /// <param name="request">request to make</param>
+        /// <exception>throws exception on network or connection errors.</exception>
         /// <returns>response of the request</returns>
-        public abstract System.Threading.Tasks.Task<Response> MakeRequestAysnc(Request request);
+        public abstract System.Threading.Tasks.Task<Response> MakeRequestAsync(Request request);
```
Comment (efossier) [inline src/Twilio/Http/HttpClient.cs]: ⚠️  This is technically backwards incompatible to anyone who has implemented their own C# client which would have to extend this class.

File: src/Twilio/Clients/TwilioRestClient.cs [hunk 8]
```diff
@@ -67,7 +72,19 @@ public class TwilioRestClient : ITwilioRestClient
         public Response Request(Request request)
         {
             request.SetAuth(_username, _password);
-            var response = HttpClient.MakeRequest(request);
+            Response response;
+            try
+            {
+                response = HttpClient.MakeRequest(request);
+            }
+            catch (TwilioException) { throw; }  /* if client throws TwilioException then let it passthrough */
```
Comment (dougblack) [inline src/Twilio/Clients/TwilioRestClient.cs]: If the new clients can't throw `TwilioException` let's remove this catch.

File: src/Twilio/Clients/TwilioRestClient.cs [hunk 9]
```diff
@@ -81,9 +98,22 @@ public Response Request(Request request)
         public async Task<Response> RequestAsync(Request request)
         {
             request.SetAuth(_username, _password);
-            var response = await HttpClient.MakeRequestAysnc(request);
+            Response response;
+            try
+            {
+                response = await HttpClient.MakeRequestAsync(request);
+            }
+            catch (TwilioException) { throw; }  /* if client throws TwilioException then let it passthrough */
```
Comment (dougblack) [inline src/Twilio/Clients/TwilioRestClient.cs]: Same here.

File: src/Twilio/Http/SystemNetHttpClient.cs (hunk 6 above)
Comment (dougblack) [inline src/Twilio/Http/SystemNetHttpClient.cs]: IMO we should always throw the first exception. We know the thing we're calling will only throw one exception so feels like a safe assumption.

File: src/Twilio/Http/Net35/WebRequestClient.cs [hunk 10]
```diff
@@ -0,0 +1 @@
...

            var authBytes = Authentication(request.Username, request.Password);
            httpRequest.Headers["Authorization"] = "Basic " + authBytes;
            httpRequest.ContentType = "application/x-www-form-urlencoded";

            return httpRequest;
        }
    }
}
#endif
```
Comment (dougblack) [inline src/Twilio/Http/Net35/WebRequestClient.cs]: What is going on here?

Comment (efossier) [inline src/Twilio/Http/Net35/WebRequestClient.cs]: damn i thought i fixed that. Its messed up line endings from working on the windows machine..

**Comment by dprothero** [comment]
LGTM 👍

**Pull request merged**